mostlyai/qa/_coherence.py

```diff
@@ -0,0 +1,196 @@
+# Copyright 2025 MOSTLY AI
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from joblib import Parallel, cpu_count, delayed, parallel_config
+import pandas as pd
+
+from mostlyai.qa._accuracy import (
+    calculate_accuracy,
+    calculate_accuracy_cnts,
+    plot_univariate,
+)
+from mostlyai.qa._filesystem import TemporaryWorkspace
+
+
+def calculate_distinct_categories_per_sequence(df: pd.DataFrame, context_key: str) -> pd.DataFrame:
+    return df.groupby(context_key).nunique().reset_index(drop=True)
+
+
+def calculate_distinct_categories_per_sequence_accuracy(
+    trn_binned_cats_per_seq: pd.DataFrame, syn_binned_cats_per_seq: pd.DataFrame
+) -> pd.DataFrame:
+    acc_cats_per_seq = pd.DataFrame({"column": trn_binned_cats_per_seq.columns})
+    with parallel_config("loky", n_jobs=min(cpu_count() - 1, 16)):
+        results = Parallel()(
+            delayed(calculate_accuracy)(
+                trn_bin_cols=trn_binned_cats_per_seq[[row["column"]]],
+                syn_bin_cols=syn_binned_cats_per_seq[[row["column"]]],
+            )
+            for _, row in acc_cats_per_seq.iterrows()
+        )
```

Comment:
We would probably boost the performance significantly by using list comprehensions (or `.apply()`). For example:
```python
        results = Parallel()(
            delayed(calculate_accuracy)(
                trn_bin_cols=trn_binned_cats_per_seq[[col]],
                syn_bin_cols=syn_binned_cats_per_seq[[col]],
            )
            for col in acc_cats_per_seq["column"]
        )
```